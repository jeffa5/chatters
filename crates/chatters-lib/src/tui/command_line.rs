use crossterm::event::KeyEvent;
use tui_textarea::TextArea;

use crate::{command_history::CommandLineHistory, commands::Completion};

use super::list::ListState;

#[derive(Debug, Default)]
pub struct CommandLine {
    command: TextArea<'static>,
    pub error: String,
    pub completions: Completions,
    pub history: CommandLineHistory,
}

impl CommandLine {
    pub fn text(&self) -> &str {
        self.command.lines().first().unwrap()
    }

    pub fn text_without_completion(&self) -> String {
        if let Some(completion) = self
            .completions
            .list_state
            .selected()
            .map(|i| self.completions.candidates[i].clone())
        {
            let mut textarea = self.command.clone();
            let char_count = completion.append.chars().count();
            for _ in 0..char_count {
                textarea.move_cursor(tui_textarea::CursorMove::Back);
            }
            textarea.delete_str(char_count);
            textarea.lines().first().unwrap().to_owned()
        } else {
            self.text().to_owned()
        }
    }

    pub fn set_text(&mut self, text: String) {
        self.command = TextArea::new(vec![text]);
        self.command.move_cursor(tui_textarea::CursorMove::End);
    }

    pub fn append_text(&mut self, text: String) {
        self.command.insert_str(text);
    }

    pub fn clear(&mut self) {
        self.command = TextArea::default();
        self.error.clear();
        self.completions.clear();
        self.history.clear_selection();
    }

    pub fn cursor_index(&self) -> usize {
        self.command.cursor().1
    }

    pub fn input(&mut self, key_event: KeyEvent) {
        self.completions.clear();
        self.command.input(key_event);
    }

    pub fn completions(&self) -> &[Completion] {
        self.completions.candidates()
    }

    pub fn set_completions(&mut self, completions: Vec<Completion>) {
        self.completions
            .set_completions(completions, self.text_without_completion());
    }

    pub fn selected_completion(&self) -> Option<usize> {
        self.completions.list_state.selected()
    }

    pub fn textarea(&mut self) -> &mut TextArea<'static> {
        &mut self.command
    }

    pub fn select_next_completion(&mut self) {
        let last = self
            .completions
            .list_state
            .selected()
            .map(|i| self.completions.candidates[i].clone());
        self.completions.select_next();
        if let Some(last) = last {
            let char_count = last.append.chars().count();
            for _ in 0..char_count {
                self.command.move_cursor(tui_textarea::CursorMove::Back);
            }
            self.command.delete_str(char_count);
        }
        if let Some(comp) = self
            .completions
            .list_state
            .selected()
            .map(|i| self.completions.candidates[i].clone())
        {
            self.command.insert_str(comp.append);
        }
    }

    pub fn completions_generated_for(&self) -> &Option<String> {
        &self.completions.generated_for
    }
}

#[derive(Debug, Default)]
pub struct Completions {
    pub candidates: Vec<Completion>,
    pub list_state: ListState,
    pub generated_for: Option<String>,
}

impl Completions {
    pub fn candidates(&self) -> &[Completion] {
        &self.candidates
    }

    pub fn set_completions(&mut self, completions: Vec<Completion>, generated_for: String) {
        if Some(&generated_for) != self.generated_for.as_ref() {
            self.list_state.select(None);
            self.candidates = completions;
            self.generated_for = Some(generated_for);
        }
    }

    pub fn clear(&mut self) {
        self.candidates.clear();
        self.list_state.select(None);
        self.generated_for = None;
    }

    fn select_next(&mut self) {
        if self.candidates.is_empty() {
            return;
        }
        if let Some(index) = self.list_state.selected() {
            let new_index = index + 1;
            if new_index == self.candidates.len() {
                self.list_state.select(None);
            } else {
                self.list_state.select(Some(new_index));
            }
        } else {
            self.list_state.select(Some(0));
        }
    }
}
