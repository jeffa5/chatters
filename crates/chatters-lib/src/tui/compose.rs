use std::{os::unix::fs::MetadataExt as _, path::PathBuf};

use crossterm::event::KeyEvent;
use ratatui::{
    layout::{Constraint, Layout},
    style::Style,
    widgets::{Block, Paragraph},
};
use tui_textarea::TextArea;

use crate::backends::MessageAttachment;

use super::messages::Quote;

#[derive(Debug, Default)]
pub struct Compose {
    textarea: TextArea<'static>,
    block: Block<'static>,
    quote: Option<Quote>,
    attachments: Vec<MessageAttachment>,
}

impl Compose {
    pub fn set_text(&mut self, lines: Vec<String>) {
        self.textarea = TextArea::new(lines);
    }

    pub fn set_quote(&mut self, quote: Quote) {
        self.quote = Some(quote);
    }

    pub fn quote(&self) -> &Option<Quote> {
        &self.quote
    }

    pub fn attachments(&self) -> &[MessageAttachment] {
        &self.attachments
    }

    pub fn attach_file(&mut self, path: PathBuf) {
        // TODO: check for duplicate path
        self.attachments.push(MessageAttachment {
            name: path.file_name().unwrap().to_string_lossy().into_owned(),
            index: 0,
            size: path.metadata().map(|m| m.size()).unwrap_or(0),
            path: Some(path),
        })
    }

    pub fn detach_file(&mut self, index: usize) {
        if index >= self.attachments.len() {
            return;
        }
        self.attachments.remove(index);
    }

    pub fn lines(&self) -> &[String] {
        self.textarea.lines()
    }

    pub fn set_block(&mut self, block: Block<'static>) {
        self.block = block;
    }

    pub fn set_cursor_style(&mut self, style: Style) {
        self.textarea.set_cursor_style(style);
    }

    pub fn set_cursor_line_style(&mut self, style: Style) {
        self.textarea.set_cursor_line_style(style);
    }

    pub fn input(&mut self, key_event: KeyEvent) {
        self.textarea.input(key_event);
    }

    pub fn clear(&mut self) {
        self.textarea = TextArea::default();
        self.quote = None;
        self.attachments.clear();
    }

    pub fn height(&self) -> u16 {
        self.quote.as_ref().map_or(0, |_| 1)
            + self.attachments.len() as u16
            + self.textarea.lines().len().max(1) as u16
            + 1
        // 1 for top border
    }
}

impl ratatui::widgets::Widget for &Compose {
    fn render(self, area: ratatui::prelude::Rect, buf: &mut ratatui::prelude::Buffer)
    where
        Self: Sized,
    {
        let mut constraints = Vec::new();

        let mut attachments_index = 0;
        let mut textarea_index = 0;

        if self.quote.is_some() {
            constraints.push(Constraint::Length(1));
            attachments_index += 1;
            textarea_index += 1;
        }
        if !self.attachments.is_empty() {
            constraints.push(Constraint::Length(self.attachments.len() as u16));
            textarea_index += self.attachments.len();
        }
        constraints.push(Constraint::Length(self.textarea.lines().len().max(1) as u16));

        self.block.clone().render(area, buf);
        let area = self.block.inner(area);
        let vertical = Layout::vertical(constraints).split(area);

        if let Some(quote) = &self.quote {
            let quote_text = quote.text.lines().next().unwrap();
            let quote_text = format!("> {quote_text}");
            Paragraph::new(quote_text).render(vertical[0], buf);
        }

        if !self.attachments.is_empty() {
            let mut lines = Vec::new();
            for attachment in &self.attachments {
                lines.push(ratatui::text::Line::from(attachment.message_line()));
            }
            Paragraph::new(lines).render(vertical[attachments_index], buf);
        }

        self.textarea.render(vertical[textarea_index], buf)
    }
}
