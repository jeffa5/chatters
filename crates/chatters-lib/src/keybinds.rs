use std::{fmt::Display, str::FromStr};

use crossterm::event::{KeyCode, KeyModifiers};

use crate::tui::Mode;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyEvent {
    pub code: KeyCode,
    pub modifiers: KeyModifiers,
}

impl Display for KeyEvent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let is_plain_char = matches!(self.code, KeyCode::Char(_)) && self.modifiers.is_empty();
        let is_capital = matches!(self.code, KeyCode::Char(c) if c.is_uppercase());
        let is_plain_capital = is_capital && self.modifiers == KeyModifiers::SHIFT;

        if is_plain_char || is_plain_capital {
            write!(f, "{}", self.code)
        } else {
            write!(f, "<")?;
            if self.modifiers.contains(KeyModifiers::CONTROL) {
                write!(f, "C-")?;
            }
            if self.modifiers.contains(KeyModifiers::SHIFT) && !is_capital {
                write!(f, "S-")?;
            }
            if self.modifiers.contains(KeyModifiers::ALT) {
                write!(f, "A-")?;
            }
            write!(f, "{}>", self.code)
        }
    }
}

impl FromStr for KeyEvent {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut slf = Self {
            code: KeyCode::Esc,
            modifiers: KeyModifiers::empty(),
        };
        if s.starts_with("<") && s.ends_with(">") {
            let mut inner = &s[1..s.len() - 1];

            loop {
                if let Some(remainder) = inner.strip_prefix("c-") {
                    slf.modifiers |= KeyModifiers::CONTROL;
                    inner = remainder;
                    continue;
                }
                if let Some(remainder) = inner.strip_prefix("C-") {
                    slf.modifiers |= KeyModifiers::CONTROL;
                    inner = remainder;
                    continue;
                }
                if let Some(remainder) = inner.strip_prefix("a-") {
                    slf.modifiers |= KeyModifiers::ALT;
                    inner = remainder;
                    continue;
                }
                if let Some(remainder) = inner.strip_prefix("A-") {
                    slf.modifiers |= KeyModifiers::ALT;
                    inner = remainder;
                    continue;
                }
                if let Some(remainder) = inner.strip_prefix("s-") {
                    slf.modifiers |= KeyModifiers::SHIFT;
                    inner = remainder;
                    continue;
                }
                if let Some(remainder) = inner.strip_prefix("S-") {
                    slf.modifiers |= KeyModifiers::SHIFT;
                    inner = remainder;
                    continue;
                }
                break;
            }

            slf.code = if inner.len() == 1 {
                let c = inner.chars().next().unwrap();
                if c.is_uppercase() {
                    slf.modifiers |= KeyModifiers::SHIFT;
                }
                KeyCode::Char(c)
            } else {
                match inner.to_lowercase().as_str() {
                    "enter" => KeyCode::Enter,
                    "esc" => KeyCode::Esc,
                    "up" => KeyCode::Up,
                    "down" => KeyCode::Down,
                    "left" => KeyCode::Left,
                    "right" => KeyCode::Right,
                    "home" => KeyCode::Home,
                    "end" => KeyCode::End,
                    "pageup" => KeyCode::PageUp,
                    "pagedown" => KeyCode::PageDown,
                    _ => return Err(()),
                }
            };
        } else if s.len() == 1 {
            let c = s.chars().next().unwrap();
            if c.is_uppercase() {
                slf.modifiers |= KeyModifiers::SHIFT;
            }
            slf.code = KeyCode::Char(c)
        } else {
            return Err(());
        }
        Ok(slf)
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct KeyEvents(pub Vec<KeyEvent>);

impl FromStr for KeyEvents {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut slf = Self(Vec::new());
        let mut ke = String::new();
        for c in s.chars() {
            if c == '<' {
                ke.push(c);
            } else if c == '>' {
                ke.push(c);
                let key_event = KeyEvent::from_str(&ke)?;
                slf.0.push(key_event);
                ke.clear();
            } else if ke.is_empty() {
                slf.0.push(KeyEvent::from_str(&c.to_string())?);
            } else {
                ke.push(c);
            }
        }
        Ok(slf)
    }
}

impl Display for KeyEvents {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for event in &self.0 {
            write!(f, "{}", event)?;
        }
        Ok(())
    }
}

#[derive(Debug)]
pub struct KeyBinds {
    pub normal_bindings: Vec<(KeyEvents, String)>,
    pub command_bindings: Vec<(KeyEvents, String)>,
    pub compose_bindings: Vec<(KeyEvents, String)>,
    pub popup_bindings: Vec<(KeyEvents, String)>,
}

impl Default for KeyBinds {
    fn default() -> Self {
        let mut normal = Vec::<(KeyEvents, String)>::new();
        normal.push((KeyEvents::from_str("q").unwrap(), ":quit<enter>".to_owned()));
        normal.push((
            KeyEvents::from_str("J").unwrap(),
            ":next-contact<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("<s-down>").unwrap(),
            ":next-contact<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("K").unwrap(),
            ":prev-contact<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("<s-up>").unwrap(),
            ":prev-contact<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("j").unwrap(),
            ":next-message<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("<down>").unwrap(),
            ":next-message<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("k").unwrap(),
            ":prev-message<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("<up>").unwrap(),
            ":prev-message<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("i").unwrap(),
            ":mode-compose<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("g").unwrap(),
            ":select-message 0<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("G").unwrap(),
            ":select-message -1<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("I").unwrap(),
            ":compose-in-editor<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("<enter>").unwrap(),
            ":send-message<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("?").unwrap(),
            ":keybindings<enter>".to_owned(),
        ));
        normal.push((
            KeyEvents::from_str("h").unwrap(),
            ":command-history<enter>".to_owned(),
        ));

        let mut popup = Vec::<(KeyEvents, String)>::new();
        popup.push((
            KeyEvents::from_str("j").unwrap(),
            ":scroll-popup 1<enter>".to_owned(),
        ));
        popup.push((
            KeyEvents::from_str("k").unwrap(),
            ":scroll-popup -1<enter>".to_owned(),
        ));

        Self {
            normal_bindings: normal,
            command_bindings: Vec::new(),
            compose_bindings: Vec::new(),
            popup_bindings: popup,
        }
    }
}

impl KeyBinds {
    pub fn get(&self, events: &KeyEvents, mode: Mode) -> Result<&String, bool> {
        let bindings = match mode {
            Mode::Normal => &self.normal_bindings,
            Mode::Command { .. } => &self.command_bindings,
            Mode::Compose => &self.compose_bindings,
            Mode::Popup => &self.popup_bindings,
        };
        let mut prefix = false;
        for (keys, command) in bindings {
            if keys == events {
                return Ok(command);
            }
            if keys.0.starts_with(&events.0) {
                prefix = true;
            }
        }
        Err(prefix)
    }

    pub fn iter(&self, mode: Mode) -> impl Iterator<Item = &(KeyEvents, String)> {
        match mode {
            Mode::Normal => &self.normal_bindings,
            Mode::Command { .. } => &self.command_bindings,
            Mode::Compose => &self.compose_bindings,
            Mode::Popup => &self.popup_bindings,
        }
        .iter()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_key_events() {
        insta::assert_debug_snapshot!(KeyEvents::from_str("a<enter><c-esc>"));
    }

    #[test]
    fn parse_key_event() {
        insta::assert_debug_snapshot!(KeyEvent::from_str("a"));
        insta::assert_debug_snapshot!(KeyEvent::from_str("<c-a>"));
        insta::assert_debug_snapshot!(KeyEvent::from_str("<s-c-a>"));
        insta::assert_debug_snapshot!(KeyEvent::from_str("<esc>"));
    }

    #[test]
    fn display_key_event() {
        insta::assert_debug_snapshot!(vec![
            KeyEvent::from_str("a").unwrap().to_string(),
            KeyEvent::from_str("A").unwrap().to_string(),
            KeyEvent::from_str("<c-a>").unwrap().to_string(),
            KeyEvent::from_str("<c-A>").unwrap().to_string(),
            KeyEvent::from_str("<up>").unwrap().to_string(),
            KeyEvent::from_str("<c-up>").unwrap().to_string(),
        ]);
    }
}
