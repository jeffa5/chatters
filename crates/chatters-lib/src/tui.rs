use command_line::CommandLine;
use compose::Compose;
use contacts::Contacts;
use log::warn;
use messages::Message;
use messages::Messages;
use ratatui::layout::Alignment;
use ratatui::layout::Constraint;
use ratatui::layout::Direction;
use ratatui::layout::Flex;
use ratatui::layout::Layout;
use ratatui::layout::Margin;
use ratatui::layout::Rect;
use ratatui::style::Style;
use ratatui::style::Styled;
use ratatui::style::Stylize;
use ratatui::text::Line;
use ratatui::text::Text;
use ratatui::widgets::Block;
use ratatui::widgets::Borders;
use ratatui::widgets::Clear;
use ratatui::widgets::List;
use ratatui::widgets::ListState;
use ratatui::widgets::Paragraph;
use ratatui::widgets::Row;
use ratatui::widgets::Scrollbar;
use ratatui::widgets::ScrollbarOrientation;
use ratatui::widgets::ScrollbarState;
use ratatui::widgets::Table;
use ratatui::Frame;
use std::fmt::Display;
use textwrap::Options;

use crate::backends::Contact;
use crate::backends::ContactId;
use crate::keybinds::KeyBinds;

mod command_line;
mod compose;
mod contacts;
mod messages;
pub use messages::Quote;

fn timestamp() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .expect("Time went backwards")
        .as_millis() as u64
}

#[derive(Debug, Clone, Copy)]
pub enum BasicMode {
    Normal,
    Popup,
    Compose,
}

#[derive(Debug, Default, Clone, Copy)]
pub enum Mode {
    #[default]
    Normal,
    Command {
        previous: BasicMode,
    },
    Compose,
    Popup,
}

impl Display for Mode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Mode::Normal => "Normal",
            Mode::Command { previous: _ } => "Command",
            Mode::Compose => "Compose",
            Mode::Popup => "Popup",
        };
        f.write_str(s)
    }
}

#[derive(Debug)]
pub struct Popup {
    pub typ: PopupType,
    pub scroll: u16,
}

impl Popup {
    pub fn new(typ: PopupType) -> Self {
        Self { typ, scroll: 0 }
    }
}

#[derive(Debug)]
pub enum PopupType {
    MessageInfo { timestamp: u64 },
    ContactInfo { id: ContactId },
    Keybinds,
    Commands,
    CommandHistory,
}

#[derive(Debug, Default)]
pub struct TuiState {
    pub self_id: Vec<u8>,
    pub attachments_path: String,
    pub message_list_state: ListState,
    pub contacts: Contacts,
    pub messages: Messages,
    pub compose: Compose,
    pub command_line: CommandLine,
    pub mode: Mode,
    pub popup: Option<Popup>,
}

impl TuiState {
    pub fn selected_message(&self) -> Option<&Message> {
        self.message_list_state
            .selected()
            .and_then(|i| self.messages.get_by_index(i))
    }
}

pub fn render(frame: &mut Frame<'_>, tui_state: &mut TuiState) {
    let now = timestamp();
    let area = frame.area();
    let vertical_splits = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Min(1),
            Constraint::Length(1),
            Constraint::Length(1),
        ])
        .split(area);

    let contacts_messages =
        Layout::horizontal([Constraint::Percentage(25), Constraint::Percentage(75)])
            .split(vertical_splits[0]);

    render_contacts(frame, contacts_messages[0], tui_state, now);

    let compose_height = tui_state.compose.height();
    let message_rect = Layout::vertical([Constraint::Fill(1), Constraint::Length(compose_height)])
        .split(contacts_messages[1]);

    render_messages(frame, message_rect[0], tui_state, now);
    render_compose(frame, message_rect[1], tui_state, now);

    render_status(frame, vertical_splits[1], tui_state, now);
    render_command(frame, vertical_splits[2], tui_state, now);

    render_popup(frame, area, tui_state);
}

fn render_contacts(frame: &mut Frame<'_>, rect: Rect, tui_state: &mut TuiState, now: u64) {
    let contact_items: Vec<_> = tui_state
        .contacts
        .iter_contacts_and_groups()
        .map(|c| {
            let age = if c.last_message_timestamp == 0 {
                String::new()
            } else {
                biggest_duration_string(now.saturating_sub(c.last_message_timestamp))
            };
            Row::new(vec![
                Text::from(c.name.to_string()),
                Text::from(age).alignment(Alignment::Right),
            ])
        })
        .collect();
    let contact_items_len = contact_items.len();
    let block = Block::new().borders(Borders::RIGHT);
    let area = block.inner(rect);
    frame.render_widget(block, rect);

    let contacts = Table::new(contact_items, [Constraint::Fill(1), Constraint::Length(3)])
        .row_highlight_style(Style::new().reversed());

    let remaining_area = render_scrollbar(
        frame,
        area,
        contact_items_len,
        tui_state.contacts.state.offset(),
    );

    frame.render_stateful_widget(contacts, remaining_area, &mut tui_state.contacts.state);
}

fn render_messages(frame: &mut Frame<'_>, rect: Rect, tui_state: &mut TuiState, now: u64) {
    let message_width = rect.width as usize;
    let message_items = tui_state.messages.messages_by_ts.values().map(|m| {
        let sender_width = 20;
        let sender = tui_state
            .contacts
            .contact_by_id(&m.sender)
            .map(|c| c.name.clone())
            .unwrap();
        let sender = truncate_or_pad(sender, sender_width);
        let age = biggest_duration_string(now.saturating_sub(m.timestamp));
        let sender_time = format!("{sender} {age:>3} ");

        let content_width = message_width
            .saturating_sub(sender_time.len())
            .saturating_sub(1);
        let content_indent = " ".repeat(sender_time.len());

        let mut lines = m.render(content_width);

        for (i, line) in lines.iter_mut().enumerate() {
            if i == 0 {
                line.insert_str(0, &sender_time);
            } else {
                line.insert_str(0, &content_indent);
            }
        }
        if lines.is_empty() {
            warn!(message:? = m; "Message with no information...");
        }
        lines.join("\n")
    });
    let messages = List::default()
        .items(message_items)
        .highlight_style(Style::new().reversed());

    let remaining_area = render_scrollbar(
        frame,
        rect,
        messages.len(),
        tui_state.message_list_state.offset(),
    );

    frame.render_stateful_widget(&messages, remaining_area, &mut tui_state.message_list_state);
}

fn render_compose(frame: &mut Frame<'_>, rect: Rect, tui_state: &mut TuiState, _now: u64) {
    tui_state
        .compose
        .set_block(Block::new().borders(Borders::TOP));
    if matches!(tui_state.mode, Mode::Compose) {
        // show cursor
        tui_state.compose.set_cursor_style(Style::new().reversed());
        tui_state
            .compose
            .set_cursor_line_style(Style::new().underlined());
    } else {
        // hide cursor
        tui_state.compose.set_cursor_style(Style::new());
        tui_state.compose.set_cursor_line_style(Style::new());
    }
    frame.render_widget(&tui_state.compose, rect);
}

fn render_status(frame: &mut Frame<'_>, rect: Rect, tui_state: &mut TuiState, _now: u64) {
    let completions = tui_state.command_line.completions().join(" ");
    let status_line = Paragraph::new(Line::from(format!("{} {}", tui_state.mode, completions)))
        .style(Style::new().reversed());
    frame.render_widget(status_line, rect);
}

fn render_command(frame: &mut Frame<'_>, rect: Rect, tui_state: &mut TuiState, _now: u64) {
    if tui_state.command_line.error.is_empty() {
        if matches!(tui_state.mode, Mode::Command { .. }) {
            let value = tui_state.command_line.text();
            (format!(":{}", value), Style::new());
            frame.render_widget(Line::from(":"), rect);
            let inner_rect = rect.inner(Margin {
                horizontal: 1,
                vertical: 0,
            });
            tui_state
                .command_line
                .textarea()
                .set_cursor_line_style(Style::new());
            frame.render_widget(&*tui_state.command_line.textarea(), inner_rect);
        } else {
            frame.render_widget(ratatui::widgets::Clear, rect);
        }
    } else {
        frame.render_widget(
            Paragraph::new(tui_state.command_line.error.clone()).set_style(Style::new().red()),
            rect,
        );
    };
}

fn biggest_duration_string(duration_ms: u64) -> String {
    let year = duration_ms / (1000 * 60 * 60 * 24 * 365);
    let month = duration_ms / (1000 * 60 * 60 * 24 * 30);
    let week = duration_ms / (1000 * 60 * 60 * 24 * 7);
    let day = duration_ms / (1000 * 60 * 60 * 24);
    let hour = duration_ms / (1000 * 60 * 60);
    let minute = duration_ms / (1000 * 60);
    let second = duration_ms / 1000;
    if year > 0 {
        format!("{year}y")
    } else if month > 0 {
        format!("{month}M")
    } else if week > 0 {
        format!("{week}w")
    } else if day > 0 {
        format!("{day}d")
    } else if hour > 0 {
        format!("{hour}h")
    } else if minute > 0 {
        format!("{minute}m")
    } else if second > 0 {
        format!("{second}s")
    } else {
        "now".to_owned()
    }
}

fn wrap_text(s: &str, width: usize) -> Text {
    let content = textwrap::wrap(s, Options::new(width))
        .into_iter()
        .map(|s| Line::from(s.into_owned()))
        .collect::<Vec<_>>();
    Text::from(content)
}

fn truncate_or_pad(mut s: String, width: usize) -> String {
    if s.len() >= width {
        s[..width].to_owned()
    } else {
        s.push_str(&" ".repeat(width - s.len()));
        s
    }
}

fn render_popup(frame: &mut Frame<'_>, area: Rect, tui_state: &mut TuiState) {
    let Some(popup) = &tui_state.popup else {
        return;
    };
    let area = popup_area(area, 60, 50);
    frame.render_widget(Clear, area); // this clears out the background
    let width = area.width.saturating_sub(2) as usize;
    let (title, text) = match &popup.typ {
        PopupType::MessageInfo { timestamp } => {
            let message = tui_state.messages.get_by_timestamp(*timestamp).unwrap();
            render_message_info(width, tui_state, message)
        }
        PopupType::ContactInfo { id } => {
            let contact = tui_state
                .contacts
                .iter_contacts_and_groups()
                .find(|c| &c.id == id)
                .unwrap();
            render_contact_info(contact)
        }
        PopupType::Keybinds => render_keybinds(),
        PopupType::Commands => render_commands(),
        PopupType::CommandHistory => render_command_line_history(tui_state),
    };

    let text = wrap_text(&text, width);

    let line_count = text.lines.len() as u16;
    let max_scroll = line_count.saturating_sub(area.height.saturating_sub(2));
    let popup = tui_state.popup.as_mut().unwrap();
    popup.scroll = popup.scroll.min(max_scroll);
    let block = Block::bordered().title(title);
    let inner_area = block.inner(area);
    frame.render_widget(block, area);

    let remaining_area =
        render_scrollbar(frame, inner_area, line_count.into(), popup.scroll.into());

    let para = Paragraph::new(text).scroll((popup.scroll, 0));
    frame.render_widget(para, remaining_area);
}

fn render_message_info(
    width: usize,
    tui_state: &TuiState,
    message: &Message,
) -> (&'static str, String) {
    let ts_seconds = message.timestamp / 1_000;
    let ts_nanos = (message.timestamp % 1_000) * 1_000_000;
    let time = chrono::DateTime::from_timestamp(
        ts_seconds.try_into().unwrap(),
        ts_nanos.try_into().unwrap(),
    )
    .unwrap();
    let sender_name = tui_state
        .contacts
        .contact_by_id(&message.sender)
        .unwrap()
        .name
        .to_string();
    let text = vec![
        format!("Sender name: {}", sender_name),
        format!("Sender id:   {:?}", message.sender),
        format!("Time:        {}", time.to_rfc3339()),
        String::new(),
        message.render(width).join("\n"),
    ]
    .join("\n");
    ("Message info", text)
}

fn render_contact_info(contact: &Contact) -> (&'static str, String) {
    let ts_seconds = contact.last_message_timestamp / 1_000;
    let ts_nanos = (contact.last_message_timestamp % 1_000) * 1_000_000;
    let time = chrono::DateTime::from_timestamp(
        ts_seconds.try_into().unwrap(),
        ts_nanos.try_into().unwrap(),
    )
    .unwrap();
    let text = vec![
        format!("Name:              {}", contact.name),
        format!("Id:                {:?}", contact.id),
        format!("Last message time: {}", time.to_rfc3339()),
        format!("Description:       {}", contact.description),
    ]
    .join("\n");
    ("Contact info", text)
}

fn render_keybinds() -> (&'static str, String) {
    let keybindings = KeyBinds::default();
    let normal_keybinds = keybindings
        .iter(Mode::Normal)
        .map(|(k, c)| format!("{} :{}", k, c.names()[0]))
        .collect::<Vec<_>>()
        .join("\n");
    let command_keybinds = keybindings
        .iter(Mode::Command {
            previous: BasicMode::Normal,
        })
        .map(|(k, c)| format!("{} :{}", k, c.names()[0]))
        .collect::<Vec<_>>()
        .join("\n");
    let compose_keybinds = keybindings
        .iter(Mode::Compose)
        .map(|(k, c)| format!("{} :{}", k, c.names()[0]))
        .collect::<Vec<_>>()
        .join("\n");
    let popup_keybinds = keybindings
        .iter(Mode::Popup)
        .map(|(k, c)| format!("{} :{}", k, c.names()[0]))
        .collect::<Vec<_>>()
        .join("\n");

    let text = format!(
        "Normal mode bindings\n{}\n\nCommand mode bindings\n{}\n\nCompose mode bindings\n{}\n\nPopup mode bindings\n{}",
        normal_keybinds, command_keybinds, compose_keybinds, popup_keybinds
    );

    ("Keybindings", text)
}

fn render_commands() -> (&'static str, String) {
    let mut commands = crate::commands::commands()
        .into_iter()
        .map(|c| {
            c.names()
                .into_iter()
                .map(|s| format!(":{s}"))
                .collect::<Vec<_>>()
                .join(", ")
        })
        .collect::<Vec<_>>();
    commands.sort();
    let text = commands.join("\n");

    ("Commands", text)
}

fn render_command_line_history(tui_state: &TuiState) -> (&'static str, String) {
    let lines = tui_state
        .command_line
        .history
        .iter()
        .into_iter()
        .map(|c| format!(":{c}"))
        .collect::<Vec<_>>();

    ("Command history", lines.join("\n"))
}

fn popup_area(area: Rect, percent_x: u16, percent_y: u16) -> Rect {
    let vertical = Layout::vertical([Constraint::Percentage(percent_y)]).flex(Flex::Center);
    let horizontal = Layout::horizontal([Constraint::Percentage(percent_x)]).flex(Flex::Center);
    let [area] = vertical.areas(area);
    let [area] = horizontal.areas(area);
    area
}

fn render_scrollbar(frame: &mut Frame<'_>, area: Rect, length: usize, position: usize) -> Rect {
    let scrollbar = Scrollbar::new(ScrollbarOrientation::VerticalRight);

    let scrollable_distance = length.saturating_sub(area.height.into());
    if scrollable_distance == 0 {
        return area;
    }

    let mut scrollbar_state = ScrollbarState::new(scrollable_distance).position(position);
    frame.render_stateful_widget(scrollbar, area, &mut scrollbar_state);

    Rect {
        width: area.width.saturating_sub(1),
        ..area
    }
}
