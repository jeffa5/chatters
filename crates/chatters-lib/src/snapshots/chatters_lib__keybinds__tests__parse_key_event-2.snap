---
source: crates/chatters-lib/src/keybinds.rs
expression: "KeyEvent::from_str(\"<c-a>\")"
---
Ok(
    KeyEvent {
        code: Char(
            'a',
        ),
        modifiers: KeyModifiers(
            CONTROL,
        ),
    },
)
